Thymesisflow software library and connections manager

mgazz opened the pull request:
libthymesisflow is the software library providing configuration capabilities for Thymesisflow.

The main features are:

1. Simple APIs to attach and detach remote memory
2. Agent keeping track of current connections
3. Command line interface to abstract low level implementation details
4. Simple interface to monitor latency and rx/tx bandwidth

More information provided in the README

Commit:
Thymesisflow software library and connections manager

christian-pinto commented on libthymesisflow/README.md:
````diff
@@ -0,0 +1,178 @@
+
+To launch unit tests, run the following command.
+
+```
+./testbin/unittest
+```
````
It is not clear that before starting the unit tests you have to start the `thymesisf-agent`. Please specify or change this behavior.

christian-pinto commented on libthymesisflow/README.md:
````diff
@@ -0,0 +1,178 @@
+make
+```
+
+**Note:** use `DEBUG=1` is required for running unittests successfully
````
Is this true for the mock build too? It looks like it is not needed when building mock libraries.

christian-pinto commented on libthymesisflow/README.md:
```diff
@@ -0,0 +1,178 @@
+
+* `rx`: number of flits read from the channel
+* `rx`: number of flits transmitted on the channel
+* `latency`: average latency on the channel
```
Is there a unit of measure for the latency metric? If yes, please specify.

christian-pinto commented on libthymesisflow/cmd/thymesisf-agent.c:
```diff
@@ -0,0 +1,150 @@
+#include "proto.h"
```
There is no copyright notice in any of the source files, please add to all files missing it (both .h and .c).

christian-pinto commented on libthymesisflow/src/thymesisflow.c:
```diff
@@ -0,0 +1,243 @@
+#include "thymesisflow.h"
+#include <fcntl.h>
+
+int detach(const uint64_t memory_size) {
```
I would consider renaming this method to something more descriptive. The name `detach` could be too generic for a code base dealing with attaching and detaching memory.

christian-pinto commented on libthymesisflow/src/thymesisflow.c:
```diff
@@ -0,0 +1,243 @@
+#ifdef MOCK
+    log_info_ext("mocking memory connection by allocating only 256 MB...\n");
+
+    if (posix_memalign((void **)&conn->ea, 128, (256 >> 20)) != 0) {
```
Defining a macro for `128` could help other developers understanding what's that value about.

mgazz commented:
Thank you @christian-pinto for the review. These are all valid points. I'll address them and update the pull requests.

Commit:
docs: add units of measures to counters

Commit:
refactor: update functions on memory hotplug/unplug

The functions `probe` and `detach` result too general
and non descriptive. Updated respectively to
`hotplug_memory_blocks` and `unplug_memory_blocks`
making a direct reference to the Memory Hotplug
functionality provided by the kernel.

Commit:
refactor: move cache alignemnt size in macro

Commit:
doc: updated build hint

Commit:
refactor: add copyright notice to source files

Commit:
refactor: isolate agent functionalities

The functionalities provided by the agent are implemented
directly in the main function. This limits reuse.
With this commit we isolate agent functionalities in a single
function called `run_agent` so that it can be leveraged in
multiple scenarios (like unit tests).

Commit:
test: automated agent management during unit tests

Remove the need of manually start the agent during unit tests.

christian-pinto commented:
All looks good to me. Just merged, thanks!
